### PR TITLE
[LLVMGPU] Add Winograd pipeline for LLVMGPU

### PR DESCRIPTION
**Pull request by Max191**
This PR adds a vectorization pass pipeline and configurations for linalg_ext.winograd ops on LLVMGPU. The pipeline follows the base vectorization pipeline, but uses scf.for thread tiling and distribution, and includes the winograd op decomposition.

### CODE REVIEW

**Comment by Max191** [comment]
Is there a place for config/pipeline tests in LLVMGPU? Should I create new files for winograd?

**Comment by qedawkins** [comment]
yeah new files would be good. Something like this: https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMGPU/test/pack_pipeline_test.mlir

and this for config: https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_matvec.mlir

**Comment by Max191** [comment]
Tests are going to fail for right now until https://github.com/iree-org/iree/pull/17200 lands

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 1]
```diff
@@ -1808,6 +1848,18 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
     LDBG("Sort Config");
     return setSortConfig(entryPointFn, sortOp, targetInfo);
   }
+  if (auto winogradOp =
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: A `TypeSwitch` would make all of this much cleaner

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 2]
```diff
@@ -245,6 +245,42 @@ void addGPUVectorizationPassPipeline(OpPassManager &funcPassManager) {
...
 
+//===---------------------------------------------------------------------===//
+// Winograd Vectorize
+//===---------------------------------------------------------------------===//
+
+void addGPUWinogradVectorizePassPipeline(OpPassManager &funcPassManager) {
+  tileAndDistributeToWorkgroup(funcPassManager);
+
+  funcPassManager.addPass(createCanonicalizerPass());
+  funcPassManager.addPass(createWorkgroupSpecializationPass());
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: Do you need the workgroup specialization pass? I am planning to deprecate it, so want to make sure we dont have spiralling uses of it.

**Review by MaheshRavishankar — COMMENTED** [review]
Ill wait for dependencies to land, but few comments.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 2 above)
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: Sorry, I forgot about this comment. I just verified performance/correctness, and I don't think this is necessary (actually seems to slightly improve benchmarks without it :) I will get rid of it.

**Review by MaheshRavishankar — APPROVED** [review]
Nice! Looks good.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 3]
```diff
@@ -962,6 +962,44 @@ static LogicalResult setFftConfig(mlir::FunctionOpInterface entryPoint,
...
+          std::is_same<WinogradOp,
+                       IREE::LinalgExt::WinogradFilterTransformOp>() ||
+          std::is_same<WinogradOp,
+                       IREE::LinalgExt::WinogradOutputTransformOp>(),
+      "expected winograd transform op");
+  auto pipeline =
+      IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUWinogradVectorize;
+  TileSizesListType tileSizes;
+  std::array<int64_t, 3> workgroupSize = {32, 4, 4};
+  auto iterationRank = op.getIterationDomainRank();
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: nit: use full type when it's not obvious based on the RHS only

**Review by kuhar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 4]
```diff
@@ -962,6 +962,44 @@ static LogicalResult setFftConfig(mlir::FunctionOpInterface entryPoint,
...
+template <typename WinogradOp>
+static LogicalResult setWinogradOpConfig(mlir::FunctionOpInterface entryPoint,
+                                         WinogradOp op,
+                                         const TargetInfo &targetInfo) {
+  static_assert(
+      std::is_same<WinogradOp, IREE::LinalgExt::WinogradInputTransformOp>() ||
+          std::is_same<WinogradOp,
+                       IREE::LinalgExt::WinogradFilterTransformOp>() ||
+          std::is_same<WinogradOp,
+                       IREE::LinalgExt::WinogradOutputTransformOp>(),
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: there should be `std::is_same_v`, no? I'm surprised this compiles

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: you can also simplify this with `llvm::is_one_of`

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:977]: OK I know why, `std::integral_constant` has a constexpr conversion operator for the value type: https://en.cppreference.com/w/cpp/types/integral_constant

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: OK I see why, you are constructing integral constants and then the conversion operator for the value type kicks in: https://en.cppreference.com/w/cpp/types/integral_constant

**Review by kuhar — APPROVED** [review]
thanks

**Pull request merged**